File: backend/app/db/models/dca_baskets.py
from __future__ import annotations
from datetime import datetime
from sqlalchemy import Column, Integer, String, Float, DateTime, Boolean, ForeignKey, JSON
from sqlalchemy.orm import relationship
from app.db.base import Base

class DCABasket(Base):
    __tablename__ = "dca_baskets"
    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False, index=True)
    name = Column(String, nullable=False)
    total_amount = Column(Float, nullable=False)
    frequency = Column(String, default="weekly")  # daily | weekly | biweekly | monthly
    day_of_week = Column(Integer, nullable=True)
    status = Column(String, default="active")  # active | paused | cancelled
    executions_count = Column(Integer, default=0)
    total_invested = Column(Float, default=0.0)
    created_at = Column(DateTime, default=datetime.utcnow)
    assets = relationship("DCABasketAsset", back_populates="basket", cascade="all, delete-orphan")

class DCABasketAsset(Base):
    __tablename__ = "dca_basket_assets"
    id = Column(Integer, primary_key=True, index=True)
    basket_id = Column(Integer, ForeignKey("dca_baskets.id", ondelete="CASCADE"), nullable=False)
    symbol = Column(String, nullable=False)
    allocation_pct = Column(Float, nullable=False)
    asset_type = Column(String, default="etf")  # stock | crypto | etf
    basket = relationship("DCABasket", back_populates="assets")

File: frontend/src/api/dcaBaskets.ts
import client from "./client";

export interface DCABasketAsset { symbol: string; allocation_pct: number; asset_type: string; }
export interface DCABasket {
  id: number; name: string; total_amount: number; frequency: string;
  status: string; executions_count: number; total_invested: number;
  created_at: string; assets: DCABasketAsset[];
}
export interface DCATemplate { name: string; total_amount: number; frequency: string; assets: DCABasketAsset[]; }

export const getBaskets = () => client.get<DCABasket[]>("/dca-baskets/").then(r => r.data);
export const getTemplates = () => client.get<DCATemplate[]>("/dca-baskets/templates").then(r => r.data);
export const createBasket = (data: { name: string; total_amount: number; frequency: string; assets: DCABasketAsset[] }) =>
  client.post<DCABasket>("/dca-baskets/", data).then(r => r.data);
export const deleteBasket = (id: number) => client.delete(`/dca-baskets/${id}`);
export const pauseBasket = (id: number) => client.post(`/dca-baskets/${id}/pause`).then(r => r.data);
export const resumeBasket = (id: number) => client.post(`/dca-baskets/${id}/resume`).then(r => r.data);
